New cop `FormatParamterMismatch` checks `format` for wrong parameter count

edmz opened the pull request:
Currently works for `format` and `sprintf` only.

jonas054 commented:
This is a good idea for a cop, but you're only halfway done. You need to:
* add tests to cover the new functionality in `spec/rubocop/cop/lint/format_parameter_mismatch.rb`
* add default configuration in `config/enabled.yml`
* add a note about the new feature in `CHANGELOG.md`
* support formatting calls that use the `%` operator
* pass RuboCop's self inspection (see the Travis CI log)

edmz commented:
I agree. @bbatsov was going to comment on it first since I have never done
a cop before.

Thanks for letting me know whsts missing.

On Tuesday, July 21, 2015, Jonas Arvidsson <notifications@github.com> wrote:

> This is a good idea for a cop, but you're only halfway done. You need to:
>
>    - add tests to cover the new functionality in
>    spec/rubocop/cop/lint/format_parameter_mismatch.rb
>    - add default configuration in config/enabled.yml
>    - add a note about the new feature in CHANGELOG.md
>    - support formatting calls that use the % operator
>    - pass RuboCop's self inspection (see the Travis CI log)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bbatsov/rubocop/pull/2057#issuecomment-123446203>.
>


-- 
Eduardo

edmz commented:
I've made some progress on the tests, but somehow, I cant run 'rake spec' without getting:

    rubocop/spec/rubocop/cop/lint/format_parameter_mismatch.rb:5:in `<top (required)>': uninitialized constant RuboCop::Cop::Lint::FormatParameterMismatch (NameError)

Any idea what could be wrong?

bbatsov commented:
Likely you didn't add it here - https://github.com/bbatsov/rubocop/blob/master/lib/rubocop.rb

Grepping for the name of some cop is a good way to find all the changes you have to make when introducing a new one.

jonas054 commented:
Yes. That's because you haven't yet required the new file in `lib/rubocop.rb`. The list there is mostly in alphabetical order, so add it in its proper place. You should name the spec file `spec/rubocop/cop/lint/format_parameter_mismatch_spec.rb`, by the way.

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,78 @@
+  module Cop
+    module Lint
+      # This lint sees if there is a mismatch between the number of
+      # expected parameters for format and what is actually passed.
```
This class comment should also include some examples of the kind of code that would trigger the cop.

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,78 @@
+      # This lint sees if there is a mismatch between the number of
+      # expected parameters for format and what is actually passed.
+      class FormatParameterMismatch < Cop
+
```
Kill this blank line.

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,78 @@
+          end
+        end
+
+        def expected_fields(args)
```
Fields or arguments?

edmz replied:
format and friends call '%s', '%a', etc a field.

bbatsov replied:
Fair enough.

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,78 @@
+          format_method?(:sprintf, node)
+        end
+
+        def percent?(node)
```
I'd call this "format_specifier" or something like this. Probably there's an ever more formal term.

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,78 @@
+          number_of_args_for_format = (args.size - 1)
+          number_of_expected_fields = expected_fields(args).size
+
+          "Number of fields (#{number_of_expected_fields}) mismatches number of parameters (#{number_of_args_for_format})."
```
You might extract this as a constant and populate it with `format`. :-)

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,78 @@
+            .scan(VALID_FIELDS_REGEX)
+        end
+
+        def missing_parameters?(node)
```
Technically speaking, those would be missing arguments. We talk about params in a definition and about args in an invocation.

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,78 @@
+          receiver, method_name, *args = *node
+
+          # commands have no explicit receiver
+          unless !receiver && method_name == name
```
`unless` with `else` should always be switched to `if/else`. Way easier to comprehend for the human mind.

edmz commented:
Thanks for the input. I am making progress but I think I hit a roadblock.

Is there a way to access to contents of a const?

This is the case I am concerned about:

    A_CONST = "%s %s"
    format(A_CONST, 1, 2)

bbatsov commented:
That's doable, but it's pretty hard. Our policy so far has been to simply ignore such cases.

edmz commented:
Ok, sounds good.

edmz commented:
Please let me know if I am missing anything.

bbatsov commented on CHANGELOG.md:
```diff
@@ -2,6 +2,9 @@
 
 ## master (unreleased)
 
+* [#2057](https://github.com/bbatsov/rubocop/pull/2057): New cop `Lint/FormatParameterMismatch` checks
```
This should be in "New features". You should add yourself to the bottom of this file.

bbatsov commented on spec/rubocop/cop/lint/format_parameter_mismatch_spec.rb:
```diff
@@ -0,0 +1,69 @@
+    expect(cop.offenses).to be_empty
+  end
+
+  A_CONST = 'One: %s Two: %i'
```
No need to define this. The parser doesn't care. :-)

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,92 @@
+      #   format('A value: %s and another: %i', a_value)
+      #
+      class FormatParameterMismatch < Cop
+        MSG = 'Number of fields (%i) mismatches number of arguments (%i).'
```
Maybe you should mention the name of the method in the message (`format` or `sprintf`). It will make it clearer I think.

edmz replied:
Good point.

bbatsov commented on lib/rubocop/cop/lint/format_parameter_mismatch.rb:
```diff
@@ -0,0 +1,92 @@
+          args.size > 1 && args.first.type != :const
+        end
+
+        def expected_fields(node)
```
I think it'd be better if this method operated just on a string. This will make it easier to comprehend. Drilling to the source is kind of distracting.

edmz replied:
You are right on that.

bbatsov commented:
Apart from my small remarks the code looks good. Before we can merge the cop you'll also have to squash all commit together and ideally run this on a few bigger projects (e.g. Rails and Rubinius). This helps us be sure there aren't edge cases we missed.

bbatsov commented:
One more thing - I don't see support for:

```ruby
puts "%.4s\\%.2s\\%s" % ["1","2","3"]
```

edmz commented:
Good catch!

On Wed, Jul 22, 2015 at 12:46 PM Bozhidar Batsov <notifications@github.com>
wrote:

> One more thing - I don't see support for:
>
> puts "%.4s\\%.2s\\%s" % ["1","2","3"]
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bbatsov/rubocop/pull/2057#issuecomment-123804921>.
>

edmz commented:
I will try to add more coverage and use cases, but it might be very hard to be exhaustive due to the flexibility of what sprintf() can receive.

bbatsov commented:
No problem. Take your time.

bbatsov commented:
@edmz Heads up - I plan to release 0.33 later this week. If you'd like this cop to get in you'll have to wrap it soon.

edmz commented:
@bbatsov so I worked on this and:

- squashed everything in 1 commit
- ran it against rubinius and no crashes
- added all the format parameters found in sprintf, except for the positional ones, but having a mismatch on those will raise an exception by ruby

bbatsov commented:
You'll need to rebase your branch on top of the current `master` branch.

edmz commented:
@bbatsov Rebased! I've never done so much git-fu before :)

bbatsov commented:
I still see a merge conflict. Guess you'll have to rebase again.

edmz commented:
Rebased again. Its the changelog. I guess you merged something before that affected that. Hopefully this time it will work.

bbatsov commented:
I've noticed a typo in your commit message.

bbatsov commented on spec/rubocop/cop/lint/format_parameter_mismatch_spec.rb:
```diff
@@ -0,0 +1,116 @@
+    expect(cop.offenses).to be_empty
+  end
+
+  it 'registers an offense for %' do
```
Please, `String#%` instead of `%` - it's more informative.

bbatsov commented on spec/rubocop/cop/lint/format_parameter_mismatch_spec.rb:
```diff
@@ -0,0 +1,116 @@
+    expect(cop.messages).to eq(msg)
+  end
+
+  it 'doesnt register an offense for % when arguments and fields match' do
```
doesn't

bbatsov commented:
FormatParamterMismatch -> FormatParameterMismatch

edmz commented:
Ah, missed that. Hopefully I got it right this time.

On Fri, Jul 31, 2015 at 11:02 AM Bozhidar Batsov <notifications@github.com>
wrote:

> FormatParamterMismatch -> FormatParameterMismatch
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/bbatsov/rubocop/pull/2057#issuecomment-126735232>.
>

bbatsov commented:
:+1: We're good to go! Thanks!
